src/sprite_renderer.c

```diff
@@ -0,0 +1,1277 @@
+
+#include <nitro.h>
+#include <string.h>
+
+#include "struct_defs/sprite_template.h"
+#include "struct_defs/struct_0200D0F4.h"
+
+#include "overlay022/struct_ov22_022559F8.h"
+#include "overlay104/struct_ov104_022412F4.h"
+#include "overlay104/struct_ov104_02241308.h"
+#include "overlay104/struct_ov104_0224133C.h"
+
+#include "cell_actor.h"
+#include "heap.h"
+#include "narc.h"
+#include "palette.h"
+#include "sprite_resource.h"
+#include "unk_020093B4.h"
+#include "unk_0200A328.h"
+#include "unk_0200A784.h"
+#include "unk_02017728.h"
+#include "unk_0201DBEC.h"
+#include "unk_0201E86C.h"
+#include "unk_0201F834.h"
+
+static BOOL SpriteRenderer_LoadResObj(SpriteRenderer *renderer, SpriteGfxHandler *gfxHandler, int narcIndex, int memberIndex, BOOL compressed, enum SpriteResourceType resourceType, int resourceID);
+static BOOL SpriteRenderer_LoadResObjFromOpenNarc(SpriteRenderer *renderer, SpriteGfxHandler *gfxHandler, NARC *narc, int memberIndex, BOOL compressed, enum SpriteResourceType resourceType, int resourceID);
+static BOOL SpriteResourceList_Add(SpriteResourceList *resourceList, SpriteResource *resource);
+static BOOL sub_0200D27C(SpriteResourceCollection *resourceCollection, SpriteResourceList *resourceList, int resourceID);
+static BOOL sub_0200D2D0(SpriteResourceCollection *resourceCollection, SpriteResourceList *resourceList, int resourceID);
+static BOOL SpriteResourceCollection_RemoveFromResourceList(SpriteResourceCollection *resourceCollection, SpriteResourceList *resourceList, int resourceID);
+
+SpriteRenderer *SpriteRenderer_Create(int heapID)
+{
+    SpriteRenderer *renderer = Heap_AllocFromHeap(heapID, sizeof(SpriteRenderer));
+
+    if (renderer == NULL) {
+        return NULL;
+    }
+
+    renderer->heapID = heapID;
+    renderer->numGfxHandlers = 0;
+    renderer->hasOamManager = TRUE;
+
+    return renderer;
+}
+
+SpriteGfxHandler *SpriteRenderer_CreateGfxHandler(SpriteRenderer *renderer)
+{
+    GF_ASSERT(renderer != NULL);
+
+    SpriteGfxHandler *gfxHandler = Heap_AllocFromHeap(renderer->heapID, sizeof(SpriteGfxHandler));
+
+    if (gfxHandler == NULL) {
+        return NULL;
+    }
+
+    renderer->numGfxHandlers++;
+
+    for (int i = 0; i < SPRITE_RESOURCE_TYPE_MAX; i++) {
+        gfxHandler->spriteResourceCollectionList[i] = NULL;
+    }
+
+    return gfxHandler;
+}
+
+GF_G2dRenderer *SpriteRenderer_GetG2dRenderer(SpriteRenderer *renderer)
+{
+    return &renderer->renderer;
+}
+
+BOOL SpriteRenderer_CreateOamCharPlttManagers(SpriteRenderer *renderer, const UnkStruct_ov104_0224133C *param1, const UnkStruct_ov104_022412F4 *param2, int param3)
+{
+    GF_ASSERT(renderer != NULL);
+
+    if (renderer == NULL) {
+        return FALSE;
+    }
+
+    {
+        UnkStruct_ov22_022559F8 v0;
+
+        v0.unk_00 = param2->unk_00;
+        v0.unk_04 = param2->unk_04;
+        v0.unk_08 = param2->unk_08;
+        v0.unk_0C = renderer->heapID;
+
+        sub_0201E88C(&v0, param2->unk_0C, param2->unk_10);
+    }
```

Comment:
**polish**: I'll leave it up to you if you want to break down this unnecessary scope-block or not; since the inner struct here is undocumented, I can let it slide.